internal/admin/snwatcher/snwatcher.go

```diff
@@ -121,60 +118,69 @@ func (snw *StorageNodeWatcher) checkHeartbeat(ctx context.Context, now time.Time
 		go func(idx int) {
 			defer wg.Done()
 			snid := md.StorageNodes[idx].StorageNodeID
-			snmd, err := snw.snmgr.GetMetadata(ctx, snid)
+
+			mctx, cancel := context.WithTimeout(ctx, snw.heartbeatCheckDeadline)
+			defer cancel()
+
+			snmd, err := snw.snmgr.GetMetadata(mctx, snid)
 			if err != nil {
+				snw.hbFail(snid)
 				errs[idx] = err
 				return
 			}
-			snw.set(snid, now)
-			snw.statsRepos.Report(ctx, snmd, now)
+			snw.hbReset(snid)
+			snw.statsRepos.Report(mctx, snmd, time.Now().UTC())
 		}(i)
 	}
 	wg.Wait()
 	return multierr.Combine(errs...)
 }
 
-func (snw *StorageNodeWatcher) set(snid types.StorageNodeID, ts time.Time) {
+func (snw *StorageNodeWatcher) hbFail(snid types.StorageNodeID) {
+	snw.mu.Lock()
+	defer snw.mu.Unlock()
+	snw.hbFailcnt[snid]++
+}
+
+func (snw *StorageNodeWatcher) hbReset(snid types.StorageNodeID) {
 	snw.mu.Lock()
 	defer snw.mu.Unlock()
-	snw.hb[snid] = ts
+	snw.hbFailcnt[snid] = 0
 }
 
 // reload resets the map lastTimes.
 // The timestamp for pre-existing storage node is retained, on the other hand,
 // the timestamp for a new one is set to the argument ts.
 // If the storage node is not in the argument snds, it is removed from the
 // lastTimes.
-func (snw *StorageNodeWatcher) reload(snds []*varlogpb.StorageNodeDescriptor, ts time.Time) {
+func (snw *StorageNodeWatcher) reload(snds []*varlogpb.StorageNodeDescriptor) {
 	snw.mu.Lock()
 	defer snw.mu.Unlock()
 
-	lastTimes := make(map[types.StorageNodeID]time.Time, len(snds))
+	lasthb := make(map[types.StorageNodeID]int, len(snds))
 	for _, snd := range snds {
 		snid := snd.StorageNodeID
-		oldts, ok := snw.hb[snid]
+		old, ok := snw.hbFailcnt[snid]
 		if ok {
-			lastTimes[snid] = oldts
+			lasthb[snid] = old
 			continue
 		}
-		lastTimes[snid] = ts
+		lasthb[snid] = 0
 	}
-	snw.hb = lastTimes
+	snw.hbFailcnt = lasthb
 }
 
-func (snw *StorageNodeWatcher) handleHeartbeatTimeout(ctx context.Context, now time.Time) {
+func (snw *StorageNodeWatcher) handleHeartbeatTimeout(ctx context.Context) {
 	snw.mu.Lock()
 	defer snw.mu.Unlock()
 
 	var wg sync.WaitGroup
-	for snid, ts := range snw.hb {
-		if now.Sub(ts) > snw.tick*time.Duration(snw.heartbeatTimeout) {
+	for snid, cnt := range snw.hbFailcnt {
+		if time.Duration(cnt)*snw.tick > snw.heartbeatTimeout {
```

Comment:
It looks good to me right now. However, the `tick` is a fixed value and the elapsed time between each tick can be greater than the fixed tick. In this case, handling heartbeat timeout might be delayed slightly. Soon, we will have to investigate the failure detection mechanism more.

Reply:
You are right. I focused on the number of heartbeat failures because our heartbeat check logic and subsequent processing logic cannot finish in a fixed amount of time.